File: chap4/QR_complete.f90
Program QR_complete
  Implicit none
  Integer :: i, j
  Integer, parameter :: m = 4, n = 2
  Real(kind=8) :: A(m,n), Q(m,m), A0(m,n), R(m,n)
  Real(kind=8) :: t1, t2
  
  a = reshape( [ 1.,2.,2.,-4.,3.,2.,2.,5. ], [m,n] )
  R = A
  call calQR( R, m, n, Q )
  Write ( *,'(1x, a)' ) 'The matrix Q is:'
  Do i = 1, m
    Write ( *,'(*(f12.5))' ) Q(i,:)
  End do
  Write ( *,'(1x, a)' ) 'The matrix R is:'
  
  Do i = 1, m
    Write ( *,'(*(f12.5))' ) R(i,:)
  End do
  
  !// 检测QR分解是否正确
  !// A = QR
  A0 = matmul( Q,R )
  Write ( *,'(1x, a)' ) 'The matrix A is:'
  Do i = 1, m
    Write ( *,'(*(f12.5))' ) A0(i,:)
  End do
  
End program QR_complete

Subroutine calQR( R, m, n, Q ) 
  Implicit none
  Real(kind=8), external :: norm_2, sign0
  Integer :: m, n, i, j
  Real(kind=8) :: R(m,n), Q(m,m), I1(m,m), Q1(m,m), U(m,1), UTU(1,1)

  Q = 0.d0
  I1 = 0.d0
  Do i = 1, m
    Q(i,i) = 1.d0
    I1(i,i) = 1.d0
  End do

  Do j = 1, n
    U = 0.d0
    Do i = j, m
      U(i,1) = R(i,j)
    End do
    U(j,1) = U(j,1) + norm_2(U,m) * sign0( U(j,1) )
    UTU = matmul( transpose(U),U )
    Q1 = I1 - 2.d0 * matmul( U,transpose(U) ) / UTU(1,1)
    R = matmul( Q1,R )
    Q = matmul( Q1,Q )
    Q = transpose( Q )
  End do

End subroutine calQR

Real(kind=8) function norm_2( U,m )
  Implicit none
  Integer m, i
  Real(kind=8) :: U(m,1)
  
  norm_2 = 0.d0
  Do i = 1, m
    norm_2 = norm_2 + U(i,1)**2
  End do
  norm_2 = sqrt(norm_2)
  
End function norm_2

Real(kind=8) function sign0(t)
  Implicit none
  Real(kind=8) :: t
  Real(kind=8), parameter :: eps = 1.d-12
  
  If ( abs(t) < eps ) then
    sign0 = 1.d0
  else
    sign0 = t / abs(t)
  End if

End function sign0

File: chap4/QR_complete.py
# -*- coding: utf-8 -*-
"""
Created on Mon May 21 21:57:07 2018

@author: luk
"""

import numpy as np
m = 4; n = 2
# A = np.array( [[1.,3.,2.], [2.,2.,3.], [2.,2.,4.], [-4.,5.,5.]], dtype = np.float64 )
A = np.array( [[1.,3.], [2.,2.], [2.,2.], [-4.,5.]], dtype = np.float64 )
A0 = np.zeros( (m,n), dtype = np.float64 )
R = np.zeros( (m,n), dtype = np.float64 )
Q = np.zeros( (m,m), dtype = np.float64 )

R[:] = A[:]

def calQR():
  global Q, R, m, n
  I1 = np.zeros( (m,m), dtype = np.float64 )
  Q1 = np.zeros( (m,m), dtype = np.float64 )
  U = np.zeros( (m,1),  dtype = np.float64 )

  for i in range(m):
    Q[i,i] = np.float64(1.0)
    I1[i,i] = np.float64(1.0)

  for j in range(n):
    U[:,0] = np.float64(0.0)
    for i in range(j,m):
      U[i,0] = R[i,j]

    U[j,0] = U[j,0] + np.sqrt( np.dot(U[:,0],U[:,0]) ) * ( U[j,0] / abs(U[j,0]) )
    UTU = np.matmul( np.transpose(U), U )
    Q1 = I1 - 2.0 * np.matmul( U, np.transpose(U) ) / UTU
    R = np.matmul( Q1, R )
    Q = np.matmul( Q1, Q )
    Q = np.transpose( Q )

calQR()

print( "The matrix Q is:" )
for i in range(m):
  print( "%9f  "*len(Q[i,:]) % tuple(Q[i,:]) )

print( "The matrix R is:" )
for i in range(m):
  print( "%9f  "*len(R[i,:]) % tuple(R[i,:]) )

# A = QR
A0 = np.matmul( Q, R )
print( "The matrix A is:" )
for i in range(m):
  print( "%9f  "*len(A0[i,:]) % tuple(A0[i,:]) )
  
print( "please input NUMBER and continue!" )
input()

print( "python求解QR" )
(QQ, RR) = np.linalg.qr( A, mode = 'complete' )
for i in range(m):
  print( "%9f  "*len(QQ[i,:]) % tuple(QQ[i,:]) )

print( "The matrix R is:" )
for i in range(m):
  print( "%9f  "*len(RR[i,:]) % tuple(RR[i,:]) )
  
AA = np.matmul( QQ, RR )
print( "The matrix A is:" )
for i in range(m):
  print( "%9f  "*len(AA[i,:]) % tuple(AA[i,:]) )
  
print( "please input NUMBER and stop!" )
input()


